modules/mod_base/filters/filter_slice.erl

```diff
@@ -38,58 +38,54 @@
 % {{ list|slice:[-3] }} -> [1,2,3,4,5,6,7]
 %
 % {{ list|slice:[M,N] }}, where N < M will return []
-% {{ list|slice:[,] }}, where N < M will return [1,2,3,4,5,6,7,8,9,0]
+% {{ list|slice:[,] }}, will return [1,2,3,4,5,6,7,8,9,0]
 %
 
 -module(filter_slice).
 -export([slice/3]).
 
-
 slice(undefined, _, _Context) ->
     undefined;
-
 slice(List, Slice, _Context) when is_list(List) ->
     slice1(List, Slice);
+slice({trans, _} = Tr, Slice, Context) ->
+    slice(z_trans:lookup_fallback(Tr, Context), Slice, Context);
 slice(Binary, Slice, _Context) when is_binary(Binary) ->
-    slice1(z_convert:to_list(Binary), Slice);
+    slice1(Binary, Slice);
 slice(MaybeList, Slice, Context) ->
     slice1(z_template_compiler_runtime:to_list(MaybeList, Context), Slice).
 
-slice1(List, [undefined, undefined]) ->
-    slice2(List, 1, length(List));
-
-slice1(List, [M, undefined]) ->
-    slice2(List, M, length(List));
-
-slice1(List, [undefined, N]) ->
-    N1 = if
-            N < 0 -> length(List) + N;
-            true -> N
-    end,
-    slice2(List, 1, N1);
-
-slice1(List, [M, N]) ->
-    slice2(List, M, N);
+slice1([], _) -> [];
+slice1(<<>>, _) -> <<>>;
+slice1(List, [undefined, undefined]) -> slice2(List, 1, strlen(List));
+slice1(List, [M, undefined]) -> slice2(List, M, strlen(List));
+slice1(List, [undefined, N]) -> slice2(List, 1, N);
+slice1(List, [M, N]) -> slice2(List, z_convert:to_integer(M), z_convert:to_integer(N));
+slice1(List, [M]) -> slice1(List, [undefined, M]);
+slice1(List, M) -> slice1(List, [undefined, z_convert:to_integer(M)]).
 
-slice1(List, [M]) ->
-    slice1(List, [undefined, M]);
+slice2(List, 0, _N) when is_list(List) -> [];
+slice2(List, _M, 0) when is_list(List) -> [];
+slice2(List, 0, _N) when is_binary(List) -> <<>>;
+slice2(List, _M, 0) when is_binary(List) -> <<>>;
+slice2(List, M, N) when M < 0 -> slice2(List, strlen(List) + M + 1, N);
+slice2(List, M, N) when N < 0 -> slice2(List, M, strlen(List)+N);
+slice2(List, M, N) when N < M, is_list(List) -> [];
+slice2(List, M, N) when N < M, is_binary(List) -> <<>>;
+slice2(List, M, N) when is_list(List) ->
+    lists:sublist(List, M, N - M + 1);
+slice2(B, M, N) when is_binary(B) ->
+    substring(B, M, N - M + 1).
 
-slice1(List, M) ->
-    slice1(List, [undefined, z_convert:to_integer(M)]).
+substring(<<>>, _, _) -> <<>>;
+substring(B, 1, N) -> substring_1(B, N, <<>>);
+substring(<<_/utf8, B/binary>>, M, N) -> substring(B, M - 1, N);
+substring(<<_, B/binary>>, M, N) -> substring(B, M - 1, N).
 
+substring_1(_, 0, Acc) -> Acc;
+substring_1(<<>>, _, Acc) -> Acc;
+substring_1(<<C/utf8, B/binary>>, N, Acc) -> substring_1(B, N - 1, <<Acc/binary, C/utf8>>);
+substring_1(<<C, B/binary>>, N, Acc) -> substring_1(B, N - 1, <<Acc/binary, C/utf8>>).
```

Comment:
Good point, I thought `binary:part` was able to handle `/utf8`, but that’s definitely not the case.